### PR TITLE
Support library test bundles

### PR DESCRIPTION
**Pull request by alanzeino**
...the library has historically been quite strict about enforcing that snapshots are run inside an Application test bundle with a Test Host. However as I discussed in the old repository, this can be relaxed for some cases:

https://github.com/facebookarchive/ios-snapshot-test-case/issues/235

### CODE REVIEW

File: docs/LibraryVsApplicationTestBundles.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+* `UIAppearance` — Most `UIAppearance` APIs break when there is no test host present.
+* `UIWindow` — You cannot make a `UIWindow` you created during your test the 'key window' because `makeKeyAndVisible` crashes at test run time. One workaround is to instead set `hidden` to `false` on the `UIWindow` instance you created. However there still won't be a 'key window' so if you have code that adds a `UIView` as a subview of the `keyWindow` then that will break.
+
+### Library tests
+Unit tests that test parts of a framework or libary should be part of a Library test bundle. This does not require a Test Host or a Simulator (though in Xcode 9, Apple still launches a Simulator for these tests).
```
Comment (bielski) [inline docs/LibraryVsApplicationTestBundles.md]: libary > library

File: FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m [hunk 2]
```diff
@@ -15,12 +15,8 @@ @implementation UIApplication (StrictKeyWindow)
 - (UIWindow *)fb_strictKeyWindow
 {
   UIWindow *keyWindow = [UIApplication sharedApplication].keyWindow;
-  if (!keyWindow) {
-    [NSException raise:@"FBSnapshotTestCaseNilKeyWindowException"
-                format:@"Snapshot tests must be hosted by an application with a key window. Please ensure your test"
-                        " host sets up a key window at launch (either via storyboards or programmatically) and doesn't"
-                        " do anything to remove it while snapshot tests are running."];
-  }
+  NSString *message = @"Snapshot tests should be hosted by an application with a key window. Please ensure your test host sets up a key window at launch (either via storyboards or programmatically) and doesn't  do anything to remove it while snapshot tests are running.";
```
Comment (krzysiekpawski) [inline FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m]: Why to log this message if `keyWindow` exists? Maybe you should keep it in `if` statement?
Also, you didn't update a comment in a header file about raising an exception.
(not sure if this method should still contain `strict` word in the name...)

File: docs/LibraryVsApplicationTestBundles.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+## Introduction
+
+Developers writing unit tests for iOS using XCTest typically use Application Test Bundles without much consideration as to _why_ their tests need to be inside an Application Test Bundle, as opposed to a Library Test Bundle. This document describes what you can and can't do with both, and why you might want to use a Library Test Bundle.
+
+Library Test Bundles were once called _Logic_ Test Bundles in Apple's nomenclature. In the context of this document, both Library and Logic are interchangeable.
+
+### Application tests
+
+Unit tests that test parts of an application (such as UIViewControllers, UIWindows, UIViews) should typically be part of an Application test bundle. An Application test bundle requires a Test Host and at test run time, a Simulator too.
```
Comment (robbertvanginkel) [inline docs/LibraryVsApplicationTestBundles.md]: Library tests also need to be run in the context of a simulator to access ios sdk specific API's. The increased functionality that the test host gives is an app context, which provides access to api's around app/viewcontroller lifecycle, functionality requiring entitlements.

File: docs/LibraryVsApplicationTestBundles.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+There are some APIs that only work inside Application test bundles. In our testing, we've seen a few. Here are some:
+
+* `-[UIControl sendActionsForControlEvents:]` — This API is commonly used to trigger actions at runtime and sometimes you might want to use it inside a test to trigger a particular code path which is ordinarily run when a user performs an action. While it does not work inside a Library test bundle, [we've written our own version for unit tests](FBSnapshotTestCase/Categories/UIControl+SendActions.h) that works well for this need. If you decide to use that category, make sure it can only be seen inside unit tests and not all of your code.
+* `UIAppearance` — Most `UIAppearance` APIs break when there is no test host present.
+* `UIWindow` — You cannot make a `UIWindow` you created during your test the 'key window' because `makeKeyAndVisible` crashes at test run time. One workaround is to instead set `hidden` to `false` on the `UIWindow` instance you created. However there still won't be a 'key window' so if you have code that adds a `UIView` as a subview of the `keyWindow` then that will break.
+
```
Comment (robbertvanginkel) [inline docs/LibraryVsApplicationTestBundles.md]: I think we also had some tests requiring keychain access that needed to be run with a test host.

**Pull request merged**